### PR TITLE
Use FLINT ball arithmetic for special functions

### PR DESCRIPTION
**Pull request by d-torrance**
We add support for FLINT's real and complex ball types to the interpreter.  (Prior to FLINT 3, they were available in the [Arb](https://www.arblib.org/) library, so we'll need to link against it in older systems.)  Eventually, it would be great if these were available at top level, maybe even as coefficients of polynomials.  But for right now, we just use them to bolster our support of special functions.

For example, many of our special functions now only take real arguments.  But with FLINT, we can support complex and real interval arguments, as well.  For example:

### Before
```m2
i1 : zeta(1/2 + ii * 14.134725141734695)
stdio:1:1:(3): error: expected a real number
```

### After
```m2
i1 : zeta(1/2 + ii * 14.134725141734695)

o1 = -1.04800845464933e-16+6.5803236815108e-16*ii

o1 : CC (of precision 53)
```

We also move several special functions that were previously handled by Boost Math over to FLINT.  There are two (`inverseRegularizedBeta` and `inverseRegularizedGamma` -- both useful for computing quantiles of important probability distributions) that don't appear to be available in FLINT, however, so we aren't able to get rid of Boost Math entirely.

This is a draft for now until I get the cmake build working and update the tests and documentation.

### CODE REVIEW

**Comment by d-torrance** [comment]
I'm going to backport FLINT 3 for the PPA so we don't need to build it for all the Ubuntu GitHub builds.  Hopefully, this fixes the weird "element not invertible" errors we're getting again (see also https://github.com/Macaulay2/M2/pull/2973#issuecomment-1778504371).

**Comment by mahrud** [comment]
That may be useful for some people, though alternatively once #3288 is working we don't need to use 22.04 for github builds.

**Comment by d-torrance** [comment]
The "element not invertible" errors are still happening with the PPA FLINT 3 package...

**Comment by mahrud** [comment]
The problem here was that ubuntu's flint was too old, and putting the newer version was incompatible with factory, right? Did you try putting newer flint together with a newer factory (and possibly normaliz) on PPA? Those are the only libraries that we use and depend on flint.

**Comment by d-torrance** [comment]
Yes, that was the problem.  I haven't tried backporting the others to Ubuntu 22.04 since I figured it would be easier to just wait for #3288.

**Comment by mahrud** [comment]
I don't know when that might happen -- I don't have a macos and I don't think Mike or anyone else has looked at it.

**Comment by d-torrance** [comment]
I suppose I could cherry-pick the Ubuntu-only changes from that PR for this one.  Would that be ok?

**Comment by mahrud** [comment]
You could do that, but I'm confused, why is M2 working on Ubuntu 22.04 not important?

**Comment by d-torrance** [comment]
It's definitely important!  I was specifically thinking about not having to spend the time building FLINT 3 during the GitHub builds.  Otherwise, in Ubuntu 22.04, we can either build FLINT 3 (cmake) or build against FLINT 2 + Arb (autotools).

**Comment by mahrud** [comment]
But that's not why this PR is failing, right? The cmake-ubuntu builds compile flint, but they still fail, and the autotools-ubuntu build also failed last time.

**Comment by d-torrance** [comment]
The autotools-ubuntu build failed last time because it was still using the FLINT 3 PPA package that caused the weird `BeginningMacaulay2` example error.  And it looks like that's what's causing the cmake builds to fail, too -- they built FLINT 3 but are using the packaged factory.

I just pushed a new version requiring Factory 4.4.0 (the first version with FLINT 3 support) in the cmake build, so hopefully that will fix those builds.

File: .github/workflows/test_build.yml [hunk 1]
```diff
@@ -160,7 +162,9 @@ jobs:
           make -C ../.. all
           export PYVERSION=`python3 -c "from sys import version_info; \
             print(f'{version_info.major}.{version_info.minor}')"`
-          export CPPFLAGS="-I`brew --prefix`/include -I`brew --prefix libomp`/include"
+          export CPPFLAGS="-I`brew --prefix`/include \
+            -I`brew --prefix libomp`/include         \
+            -I`brew --prefix flint`/include"
```
Comment (mahrud) [inline .github/workflows/test_build.yml:167]: Why is this necessary?

Comment (d-torrance) [reply .github/workflows/test_build.yml:167]: I forget.  I'll try removing it and see what happens.

File: .github/workflows/test_build.yml [hunk 2]
```diff
@@ -85,7 +85,7 @@ jobs:
           sudo apt-get install -y -q --no-install-recommends libboost-stacktrace-dev \
                   libncurses-dev libncurses5-dev libreadline-dev libeigen3-dev liblapack-dev libxml2-dev \
                   libgc-dev libgdbm-dev libglpk-dev libgmp3-dev libgtest-dev libmpfr-dev libmpfi-dev libntl-dev gfan \
-                  libgivaro-dev libboost-regex-dev fflas-ffpack libflint-dev libmps-dev libfrobby-dev \
+                  libgivaro-dev libboost-regex-dev fflas-ffpack libflint-dev libflint-arb-dev libmps-dev libfrobby-dev \
```
Comment (mahrud) [inline .github/workflows/test_build.yml:88]: Why not just resign to using the ppa package for flint and factory?

File: .github/workflows/test_build.yml [hunk 3]
```diff
@@ -72,6 +72,8 @@ jobs:
           brew tap macaulay2/tap
           brew install automake boost tbb ccache ctags llvm make ninja yasm libffi
           brew install --only-dependencies macaulay2/tap/M2
+          HOMEBREW_NO_AUTOREMOVE=1 brew uninstall factory
+          brew install singular
```
[the remark this answers is not in the document]
Comment (mahrud) [reply .github/workflows/test_build.yml]: (It is updated now but is keg-only to avoid conflicts with singular, so cp this: 59797224ce00b3be6470669cdd2b3219a9fe7f68

File: .github/workflows/test_build.yml (hunk 2 above)
Comment (d-torrance) [reply .github/workflows/test_build.yml:89]: Mostly laziness.  :)

I deleted the FLINT package from the PPA because of the `BeginningMacaulay2` example bug.  It would likely be fixed by packaging factory for the PPA, but I haven't done that.  I figure that we'll likely be switching to Ubuntu 24.04  soon enough that I'm not sure if it's worth the effort.

**Comment by mahrud** [comment]
Once #3288 is merged, could you rebase this just to confirm that the actions don't build flint? Then I think we can merge this.

**Comment by d-torrance** [comment]
> Once #3288 is merged, could you rebase this just to confirm that the actions don't build flint? Then I think we can merge this.

Sure thing!

**Comment by d-torrance** [comment]
I've rebased this onto `development` and none of the actions build flint!
